### PR TITLE
Fix #8250: Allow Expr.betaReduce to drop type ascriptions.

### LINKED ISSUE #8250: Expr.betaReduce can't reduce past the type ascription from inline def expansion
## minimized code

Macros_1.scala
```Scala
import scala.quoted._

object Macros {
  inline def betaReduce[Arg,Result](inline fn: Arg=>Result)(inline arg: Arg): Result =
    ${ betaReduceImpl('{ fn })('{ arg }) }
  def betaReduceImpl[Arg,Result](fn: Expr[Arg=>Result])(arg: Expr[Arg]): Expr[Result] =
    Expr.betaReduce(fn)(arg)
}
```

Test_2.scala
```Scala
import scala.compiletime._

object Test {
  inline def dummy: Int=>Int =
    (i: Int) => i + 1
  def main(argv: Array[String]): Unit = {
    println(code"${ Macros.betaReduce(dummy)(3) }")
  }
}
```

## Output

```Scala
(
  {
    def $anonfun(i: Int): Int = i.+(1)
    closure($anonfun)
  }
:Function1[Int, Int]).apply(
  {
    3
  }
)
```

## expectation

```Scala
4
```

### PR DESCRIPTION
**Pull request by fhackett**
Review by @nicolasstucki.

When a blackbox inline def is inlined, a type ascription is added to preserve the apparent return type of the inline def. Prior to this commit, the `quoted.Expr.betaReduce` function was unable to properly inline a known lambda expression coming from an inline def as it would see the type ascription and give up.

This change instead makes betaReduce drop type ascriptions, allowing these two features to interact properly.

Testing is done by verifying that betaReduce can correctly reduce the application of the result of an inline def to a constant value, with the reduction result being printed using the `compiletime.code` interpolator.

Consideration: dropping type ascriptions might change observable typing behaviour in some edge cases - it certainly changes one of the existing test results (see diff below). While you will certainly get _some_ type ascription back as this code can only be run from the body of an inline method (which will put back an ascription at the end), I'm not sure if that's good enough.

Since the proposed change is so small I figured I would start a discussion by submitting this PR.

### CODE REVIEW

**Comment by sjrd** [comment]
> Consideration: dropping type ascriptions might change observable typing behaviour in some edge cases - it certainly changes one of the existing test results (see diff below). While you will certainly get some type ascription back as this code can only be run from the body of an inline method (which will put back an ascription at the end), I'm not sure if that's good enough.

The crucial question is: can it change the *elaboration* of code calling that inline def? If yes (even in one edge case), then it is not acceptable.

**Comment by nicolasstucki** [comment]
There is one type ascription that should critically not be dropped. But that one is around the code inserted by the macro expansion. The fire this should not affect it. I will double check it.

File: tests/run-macros/beta-reduce-inline-result/Test_2.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+object Test {
+  
+  inline def dummy: Int => Int =
+    (i: Int) => i + 1
+
+  def main(argv : Array[String]) : Unit = {
+    println(code"${Macros.betaReduce(dummy)(3)}")
+  }
+}
+
```
Comment (nicolasstucki) [inline tests/run-macros/beta-reduce-inline-result/Test_2.scala]: We could have a simple neg test `tests/neg-macros/beta-reduce-inline-result/Test_2.scala` (using a copy of `beta-reduce-inline-result/Macro_1.scala`) to check that the elaboration is not changed.

```scala
import scala.compiletime._

object Test {

  inline def dummy: Int => Int =
    (i: Int) => i + 1
  
  val x1: Int = Macros.betaReduce(dummy)(3)
  val x2: 4   = Macros.betaReduce(dummy)(3) // error
}
```

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 2]
```diff
@@ -2043,6 +2043,8 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
     val argVals = argVals0.reverse
     val argRefs = argRefs0.reverse
     def rec(fn: Tree): Tree = fn match {
+      case Typed(expr, tpt) =>
+        rec(expr)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: We should probably get the return type out from `tpt` and add it as an ascription to the result. Even it the other tests pass, this would preserve the original typing of the inner expressions.

File: tests/run-macros/beta-reduce-inline-result/Test_2.scala (hunk 1 above)
Comment (fhackett) [reply tests/run-macros/beta-reduce-inline-result/Test_2.scala]: Agreed, I will add this.

**Comment by LPTK** [comment]
In Scala, in general you can't remove type ascriptions and expect the code to still type check.

Here is a small variation on your test:

```scala
import scala.quoted._

object Macros {
  inline def betaReduce[Arg](inline fn : Arg=>Int)(inline arg: Arg): Int =
    ${ betaReduceImpl('{ fn })('{ arg }) }

  def betaReduceImpl[Arg](fn: Expr[Arg=>Int])(arg: Expr[Arg])(using qctx : QuoteContext): Expr[Int] =
    '{${Expr.betaReduce(fn)(arg)} + 1}
}
```

```scala
import scala.compiletime._

object Test {

  inline def dummy: Int => Int =
    (i: Int) => ???

  def main(argv : Array[String]) : Unit = {
    println(code"${Macros.betaReduce(dummy)(3)}")
  }
}
```

It results in:

```scala
-- Error: tests/run-macros/beta-reduce-inline-result/Test_2.scala:9:43 -------------------------------------------------
9 |    println(code"${Macros.betaReduce(dummy)(3)}")
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |            undefined: {
  |              {
  |                {
  |                  {
  |                    ???
  |                  }
  |                }
  |              }
  |            }.+ # 10336: TermRef(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),+) at typer
  | This location contains code that was inlined from Macro_1.scala:9
```

**Comment by fhackett** [comment]
Thanks for the counter-examples. Clearly we need to replace the type ascription rather than remove it entirely - I'm currently trying to figure out how to compute the correct type to put back. Since we can't expect the ascribed type tree to be _exactly_ `AppliedType(Functionx, List(arg1, ..., result))`, I've been looking up compiler internals I can use to reliably get the right type even when type abstractions are involved.

At the moment the strategy I'll be testing when I get back to coding this is to project out the `tpt`'s apply method's `MethodType`, then catch the return type from there.

I'll also look for some more test cases with unusual ascribed type trees that refer to a function that can be beta-reduced.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala (hunk 2 above)
Comment (fhackett) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Yes, that sounds good. See my comment below for more thoughts (which I'll try when I get back to coding this).

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: It might be somthing like
```scala
Typed(rec(expr), TypeTree(tpt.tpe.finalResult))
```

**Comment by nicolasstucki** [comment]
Don't mind to much the member selection issue. Though adding the type ascription back will probably fix it. That is a more general issue (#7110).

**Comment by LPTK** [comment]
@nicolasstucki is it really a special case of #7110? Solving #7110 would not necessarily solve this. Note that field selections are not the only things that can break when dropping type ascriptions. Another example is ascription to regain a transitive subtyping relationship that's otherwise not known to Dotty.

@fhackett you should also watch out for/test with dependent function types.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala (hunk 2 above)
Comment (fhackett) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Thanks for the hint - I did not know about `finalResult`.

**Comment by fhackett** [comment]
@LPTK I'll make sure to include tests for that.

**Comment by fhackett** [comment]
@nicolasstucki I actually had no idea about the member selection issue, I just didn't know about `Type.finalResult`, which I will try rather than the construction with `apply` that I mentioned before. Interesting to learn about the other issue though.

File: tests/run-macros/beta-reduce-inline-result/Test_2.scala (hunk 1 above)
Comment (fhackett) [reply tests/run-macros/beta-reduce-inline-result/Test_2.scala]: Funny detail: this test actually passes already because the inline expansion adds its own type ascription. I'll look for a test that properly exercises the case where betaReduce's own type ascription is needed...

Comment (fhackett) [reply tests/run-macros/beta-reduce-inline-result/Test_2.scala]: ... actually @LPTK 's test should handle this. Thanks @LPTK :+1:

**Comment by fhackett** [comment]
Here is a new version taking all we've discussed so far (and some other fun details/edge cases) into account.

**Comment by nicolasstucki** [comment]
Thank you @fhackett

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 3]
```diff
@@ -2042,24 +2042,44 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
-        new TreeTypeMap(
+        val result = new TreeTypeMap(
           oldOwners = ddef.symbol :: Nil,
           newOwners = ctx.owner :: Nil,
           treeMap = tree => paramToVals.get(tree.symbol).map(_.withSpan(tree.span)).getOrElse(tree)
         ).transform(ddef.rhs)
+        topAscription match {
+          case Some(tpt) =>
+            // we assume the ascribed type has an apply that has a MethodType with a single param list (there should be no polys)
+            val methodType = tpt.tpe.member(nme.apply).info.asInstanceOf[MethodType]
```
Comment (LPTK) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: @fhackett won't this crash if the ascribed type has a `val apply` member that's not a method? I think this could happen if the term passed as `fn` is not really a function, or if the `apply` is overloaded. Something like:

```scala
scala> object test extends (String => String) { val apply = (x:Int) => x; def apply(x: String) = x }
// defined object test

scala> test.apply("abc")
val res0: String = abc

scala> test.apply
val res1: Int => Int = test$$$Lambda$1268/578362229@6f3bd37f
```

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: This case is only reached if the prefix is an explicit closure.

Comment (LPTK) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Right, I had missed the `closureDef` pattern. But it seems even a `Closure` can have a type that contains an overloaded `apply` member: indeed it [can be any SAM type](https://github.com/lampepfl/dotty/blob/a29178ce4ac848e773423f57f45186373f410970/compiler/src/dotty/tools/dotc/ast/Trees.scala#L508-L510), no?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Maybe, I will have to check if we can get one of those here. @LPTK did you have a specific example in mind?

Comment (LPTK) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Something like `betaReduce('{ (x => x) : Test }, ...)` where we have:

```scala
abstract class Test extends (String => String) { val apply = 123 }.
```

Comment (fhackett) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: I tried it, and we get a class cast exception.
I can figure out a fix - should I push a second commit here or open a second issue+PR?

Comment (fhackett) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Actually, I can do one "better". See this fun edge case:
```
abstract class Test extends (Int => Int) {
  def impl(i: Int): Int
  def apply(i: Int): Int = ???
}
```

Comment (fhackett) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: This is getting more involved than just a quick fix, so I'll open a fresh issue.

Comment (fhackett) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Continued on #8290
